### PR TITLE
Text classification: `ClassificationAdapterTask`, new `DataSources`, trainable tokenizers, allow non-Hugging Face backbones

### PR DESCRIPTION
**Pull request by pietrolesci**
## What does this PR do?

It makes the `TextClassifier` a `ClassificationAdapterTask` and decouples it from HuggingFace.

#### Model:
- [x] make `TextClassifier` an `ClassificationAdapterTask` alike what is currently done for `ImageClassifier`
- [x] implement a default adapter task
- [x] decouple `TextClassifier` from Hugging Face backend; now it does not depend on the HF keys (i.e. "input_ids" and "attention_mask")

#### Backbones:
- [x] implement HuggingFace backbone wrapper that allows to
    - [x]  choose 3 different sentence representations: `pooler_output`, `cls_token`, `avg` (of the last layer representations) - can be choosen by passing them as `backbone_kwargs`. These are implemented using FlashRegistry
    - [x] add transformers as available backbone in the `TEXT_CLASSIFIERS_BACKBONES` FlashRegistry

#### Tokenizers:
- [x] wrap Hugging Face tokenizers allowing to easily train from scratch
- [x] create a `TEXT_CLASSIFIER_TOKENIZERS` FlashRegistry

#### Data Loading:
- [x] tokenizer is now owned by the `TextClassificationPreprocess`
- [x] make `load_data` always return an Hugging Face dataset where only the labels are encoded
- [x] allow tokenizer training from scratch by hooking into the `generate_dataset` method of the `TextDataSource`
- [x] make tokenization happen in the `collate` function for maximum efficiency
    - tokenizers are fast on batches
    - dynamic padding is now possible (pad to the longest sentence in the batch, not to the maximum length)
    - does not force the user to wait until tokenization is complete every time a datamodule is created
    - normalization of the keys: the dataset only contains `DefaultDataKeys.INPUT` and `DefaultDataKeys.TARGET` (no Hugging Face-specific nomenclature)
- [x] refactor data sources code
- [x] add parquet data source 
- [x] add hugging face dataset data source 

----

Still needed:
- [ ] fix predict
- [ ] tests
- [ ] docs
- [ ] fix the pickling problem to avoid setting states all over the data.py module
- [ ] check labelstudio integration

### CODE REVIEW

File: flash/text/classification/adapters.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        self.head = head
+
+    def forward(self, x):
+        x = self.backbone(x)
+        if self.head is None:
+            return x
+        return self.head(x)
+
+
+class DefaultAdapter(Adapter):
```
Comment (tchaton) [inline flash/text/classification/adapters.py:70]: We might want to share this among data type instead.

File: flash/text/classification/backbones/__init__.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+if _TRANSFORMERS_AVAILABLE:
+    HUGGINGFACE_TEXT_CLASSIFIER_BACKBONES = ExternalRegistry(
+        getter=_trasformer,
+        name="trasformer",
+        providers=_HUGGINGFACE,
+    )
+    TEXT_CLASSIFIER_BACKBONES += HUGGINGFACE_TEXT_CLASSIFIER_BACKBONES
+
+
+# if __name__ == "__main__":
```
Comment (tchaton) [inline flash/text/classification/backbones/__init__.py:41]: Let's create tests instead.

File: flash/text/classification/backbones/transfomers.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+class Transformer(nn.Module):
+
+    available_strategies: FlashRegistry = AVAILABLE_STRATEGIES
+
+    def __init__(self, model_name: str = "prajjwal1/bert-tiny", pretrained: bool = True, strategy: str = "cls_token"):
+        super().__init__()
+        self.model_name = model_name
+        self.strategy = strategy
+        self.config = AutoConfig.from_pretrained("prajjwal1/bert-tiny")
```
Comment (tchaton) [inline flash/text/classification/backbones/transfomers.py:74]:
```suggestion
        self.config = AutoConfig.from_pretrained(model_name)
```

File: flash/text/classification/backbones/transfomers.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            self.model = AutoModel.from_config(self.config)
+
+        self._sentence_representation = AVAILABLE_STRATEGIES.get(strategy)
+
+    def forward(self, x):
+        x = self.model(x["input_ids"], x["attention_mask"])
+        return self._sentence_representation(x)
+
+
+def _trasformer(
```
Comment (tchaton) [inline flash/text/classification/backbones/transfomers.py:88]:
```suggestion
def _transformer(
```

File: flash/text/classification/backbones/transfomers.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    model_name: str = "prajjwal1/bert-tiny",
+    pretrained: bool = True,
+    strategy: str = "cls_token",
+):
+    # disable HF thousand warnings when loading model
```
Comment (tchaton) [inline flash/text/classification/backbones/transfomers.py:93]:
```suggestion
    # disable HF warnings when loading model
```

File: flash/text/classification/backbones/transfomers.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    transformers.logging.set_verbosity_error()
+
+    model = Transformer(model_name, pretrained, strategy)
+
+    # re-enable HF warnings
+    transformers.logging.set_verbosity_warning()
+
+    # # TODO: do we really need this?
```
Comment (tchaton) [inline flash/text/classification/backbones/transfomers.py:101]: Can we remove this ?

File: flash/text/classification/backbones/transfomers.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    # os.environ["TOKENIZERS_PARALLELISM"] = "TRUE"
+    # # disable HF thousand warnings
+    # warnings.simplefilter("ignore")
+    # # set os environ variable for multiprocesses
+    # os.environ["PYTHONWARNINGS"] = "ignore"
+
+    return model, model.config.hidden_size
+
+
+# if __name__ == "__main__":
```
Comment (tchaton) [inline flash/text/classification/backbones/transfomers.py:111]: Let's create test instead.

File: flash/text/classification/tokenizers/transfomers.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+# found in self-supervised learning model weights
+from typing import Generator, List, Tuple, Union
+
+import datasets
+import torch
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from transformers import AutoConfig, AutoTokenizer
+
+
+class TrasformerTokenizer:
```
Comment (tchaton) [inline flash/text/classification/tokenizers/transfomers.py:27]:
```suggestion
class TransformerTokenizer:
```

File: flash/text/classification/backbones/__init__.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+from flash.core.registry import ExternalRegistry, FlashRegistry
+from flash.core.utilities.imports import _TRANSFORMERS_AVAILABLE
+from flash.core.utilities.providers import _HUGGINGFACE
+from flash.text.classification.backbones.transfomers import _trasformer
+
+TEXT_CLASSIFIER_BACKBONES = FlashRegistry("backbones")
+
+if _TRANSFORMERS_AVAILABLE:
+    HUGGINGFACE_TEXT_CLASSIFIER_BACKBONES = ExternalRegistry(
+        getter=_trasformer,
```
Comment (tchaton) [inline flash/text/classification/backbones/__init__.py:34]: typos.

File: flash/text/classification/data.py [hunk 10]
```diff
@@ -29,17 +29,19 @@
...
 class TextDeserializer(Deserializer):
     @requires("text")
-    def __init__(self, backbone: str, max_length: int, use_fast: bool = True, **kwargs):
+    def __init__(self, backbone: str, **backbone_kwargs):
         super().__init__()
         self.backbone = backbone
-        self.tokenizer = AutoTokenizer.from_pretrained(backbone, use_fast=use_fast, **kwargs)
-        self.max_length = max_length
+        self.backbone_kwargs = backbone_kwargs
+        self.tokenizer, self.vocab = TEXT_CLASSIFIER_TOKENIZERS.get(backbone)(**backbone_kwargs)
```
Comment (tchaton) [inline flash/text/classification/data.py:41]: Neat !

**Review by tchaton — COMMENTED** [review]
Great start ! It is better practice to build the tests from the beginning. It will make you win some time.

File: flash/text/classification/backbones/__init__.py (hunk 2 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/__init__.py]: Sure, they were here for my convenience as this is still in draft. Removed.

File: flash/text/classification/backbones/transfomers.py (hunk 3 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/transfomers.py]: Done

File: flash/text/classification/backbones/transfomers.py (hunk 4 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/transfomers.py:80]: Done

File: flash/text/classification/backbones/transfomers.py (hunk 6 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/transfomers.py]: Done

File: flash/text/classification/backbones/transfomers.py (hunk 5 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/transfomers.py]: Done

File: flash/text/classification/tokenizers/transfomers.py (hunk 8 above)
Comment (pietrolesci) [reply flash/text/classification/tokenizers/transfomers.py]: Done

File: flash/text/classification/backbones/__init__.py (hunk 9 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/__init__.py]: Done

File: flash/text/classification/backbones/transfomers.py (hunk 7 above)
Comment (pietrolesci) [reply flash/text/classification/backbones/transfomers.py]: Will do
